clusterversion: significantly rework cluster version handling

irfansharif opened the pull request:
We split off ClusterVersion out of pkg/settings/cluster into a dedicated
pkg/clusterversion. This is to pave the way for #39182 where we
introduce long running version migration infrastructure.

We then introduce clusterversion.Handle as a read only view to the
active cluster version and this binary's version details. We similarly
fold in the actual global cluster version setting into
pkg/clusterversion, and enforce all external usage to go through
clusterversion.Handle. We can also hide away external usage of the baked
in cluster.Binary{,MinimumSupported}Version constants. Instead we have
the implementation of clusterversion.Handle allow for tests to override
binary and minimum supported versions as needed.

Along the way we clean up Settings.Initialized, as it is not really
used. We document all the various "versions" in play ("binary version",
    "minimum supported version", "active version") and change usage of what
was previously "serverVersion" to simply be "binaryVersion", because
that's what it is. We also clean up the Settings constructors into
Test/Non-Test types that differ in cluster version setting
initialization behaviour.

---

For reviewers: It's probably easiest to start with what 
pkg/settings/cluster/cluster_settings.go looks like, then following into
pkg/clusterversion/cluster_version.go and then pkg/clusterversion/setting.go.

---

I still don't like the following detail about our pkg structure:

- pkg/settings/cluster depends on it's "parent" pkg, pkg/settings
- pkg/settings/cluster depends pkg/clusterversion, which in turn depends
on pkg/settings

Naively, pkg/settings/cluster should be a top level package, but I'm not
making that change in this PR. For now I've renamed the settings.go file
to cluster_settings.go.

Release note: None

danhhz committed:
clusterversion: split ClusterVersion out of pkg/settings/cluster

We split off ClusterVersion out of pkg/settings/cluster into a dedicated
pkg/clusterversion. This is to pave the way for #39182 where we
introduce long running version migration infrastructure.

Release note: None

Co-authored-by: Daniel Harrison <daniel.harrison@gmail.com>
Co-authored-by: irfan sharif <irfanmahmoudsharif@gmail.com>

irfansharif committed:
roachpb: separate out comment re: minor/major version fields

https://github.com/cockroachdb/cockroach/pull/17650 renamed these fields
for the protobuf-generated C++ code, leaving the Go code unchanged.
Isolate the associated comment so it doesn't confusingly appear in the
Go code.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45455)
<!-- Reviewable:end -->

danhhz reviewed (commented):
Overall, this looks great. The way you have the constructors here is a decided improvement over my WIP. I left a couple initial nits, but I did have a hard time reviewing the file moves because it was unclear what changed. Could you please introduce a prefix commit that does only the `git mv`s so it's clear in (what will be) the second commit what changed about them?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-:-M11marI2gk7qPx7D3Q6:b3p4ty0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz, @irfansharif, @otan, and @tbg)

___
*[pkg/clusterversion/cluster_version.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M11jp4kCYl5hqegvUXK:-M11jp4kCYl5hqegvUXL:b-yx6ohs) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

nit: I'm not sure anyone knows how these dates are supposed to work, but this file is new enough content that I'd have put this as 2020.
___
*[pkg/clusterversion/cluster_version.go, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M11isAjEHVTyrnRGue4:-M11isAjEHVTyrnRGue5:b-bjyrad) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L93)):*
> ```Go
> 	// of the nodes in the cluster have running binaries that are at least as
> 	// new as that version, and that you're guaranteed that those nodes will
> 	// never be downgraded to an older version.
> ```

nit: never be downgraded to a binary with a version older than this one.
___
*[pkg/clusterversion/cluster_version.go, line 184 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M11jQ4GEuWWwuN-QmDw:-M11jQ4GEuWWwuN-QmDx:b7126is) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L184)):*
> ```Go
> 
> func (v *handleImpl) BinaryVersion() roachpb.Version {
> 	if v.binaryVersionOverride != (roachpb.Version{}) {
> ```

The MakeVersionHandle constructor actually sets this override to binaryVersion, so we'll always take this branch. That's a little odd
___
*[pkg/sql/logictest/logic.go, line 1107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M11lmNHDQ1P5NZeVj_b:-M11lmNHDQ1P5NZeVj_c:bmb2n6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/sql/logictest/logic.go#L1107)):*
> ```Go
> 			nodeParams := params.ServerArgs
> 			nodeParams.Settings = cluster.MakeTestingClusterSettingsWithVersions(
> 				cfg.binaryVersion, binaryMinSupportedVersion /* initialize */, false)
> ```

a bunch of these calls have the `/* initialize */` in the wrong place


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> pkg/settings/cluster depends pkg/clusterversion, which in turn depends
> on pkg/settings/cluster

Also, I think one of these package names should be something else, no? Go won't allow this dep graph

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-:-M11nEMbCHwRZQEgWMme:b-pehfw0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz, @irfansharif, @otan, and @tbg)



<!-- Sent from Reviewable.io -->

otan reviewed (commented):
i've also noticed `clusterversion.ClusterVersion{}` is a bit tautological. 
should we have any tests for clusterversion.go? We added some new stuff, may be worth unit testing handleImpl.
maybe worth refactoring later (but can't think of a good name)

> Reviewed 3 of 67 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-:-M123Xo2BowFfB8apJ5C:bsc6cym)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @otan, and @tbg)

___
*[pkg/clusterversion/cluster_version.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M122Knt2fpgrC_k7iCX:-M122Knt2fpgrC_k7iCY:bt2x93u) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

should this also be called `clusterversion.go` instead?
___
*[pkg/clusterversion/cluster_version.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M122SXM1xbtRT9D2SEF:-M122SXM1xbtRT9D2SEG:bplmq09) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L44)):*
> ```Go
> import (
> 	"context"
> 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
> ```

nit: it smells like this import should be in the next paragraph?
___
*[pkg/settings/cluster/cluster_settings.go, line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M123EumC5CbRn9UWVj9:-M123EumC5CbRn9UWVjA:bocjyxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/settings/cluster/cluster_settings.go#L128)):*
> ```Go
> func MakeTestingClusterSettingsWithVersions(
> 	binaryVersion, binaryMinSupportedVersion roachpb.Version,
> 	initialize bool,
> ```

nit: can we make this an enum or something to be more descriptive?


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> Could you please introduce a prefix commit that does only the git mvs so it's clear in (what will be) the second commit what changed about them?

Done.

> Also, I think one of these package names should be something else, no? Go won't allow this dep graph

Woops, fixed.

> i've also noticed clusterversion.ClusterVersion{} is a bit tautological.

Agreed, it should probably be called `pkg/version` instead. But I'll leave the restructuring for when we have a clearer idea of what it all should look like.

> We added some new stuff, may be worth unit testing handleImpl.

I think I'll hold off on this for the same reason + the fact that it's a very thin read-only shim around the global cluster version setting singleton. As I understand what pkg/clusterversion (or whatever it ends up being called) should look like, we can test the parts that make sense.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-:-M12ADnl5mh4BHEKemNm:baocpbi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @tbg)

___
*[pkg/clusterversion/cluster_version.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M11jp4kCYl5hqegvUXK:-M1279zg-F_xgkx4fG02:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L1)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: I'm not sure anyone knows how these dates are supposed to work, but this file is new enough content that I'd have put this as 2020.
</blockquote></details>

Done.
___
*[pkg/clusterversion/cluster_version.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M122Knt2fpgrC_k7iCX:-M12BSTR3H8n8xSZEcE3:b-fo2bmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L1)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

should this also be called `clusterversion.go` instead?
</blockquote></details>

Done (we don't seem to be too principled about this generally).
___
*[pkg/clusterversion/cluster_version.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M122SXM1xbtRT9D2SEF:-M12BK7L5y0vu7naFNwG:blxopgx) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L44)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: it smells like this import should be in the next paragraph?
</blockquote></details>

Woops, my linter should've caught that. Done.
___
*[pkg/clusterversion/cluster_version.go, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M11isAjEHVTyrnRGue4:-M127tDF3QsVSdyvSTFq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L93)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: never be downgraded to a binary with a version older than this one.
</blockquote></details>

Done.
___
*[pkg/clusterversion/cluster_version.go, line 184 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M11jQ4GEuWWwuN-QmDw:-M128wTqBhz1TRUOKnzw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/clusterversion/cluster_version.go#L184)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

The MakeVersionHandle constructor actually sets this override to binaryVersion, so we'll always take this branch. That's a little odd
</blockquote></details>

Done.
___
*[pkg/settings/cluster/cluster_settings.go, line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M123EumC5CbRn9UWVj9:-M12A8PX-XiM7yB_vxmq:bvxqz81) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/settings/cluster/cluster_settings.go#L128)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: can we make this an enum or something to be more descriptive?
</blockquote></details>

Done (renamed to initializeVersion).
___
*[pkg/sql/logictest/logic.go, line 1107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-M11lmNHDQ1P5NZeVj_b:-M129Vx363MOq2NVH0FO:bpxrgta) ([raw file](https://github.com/cockroachdb/cockroach/blob/57088fbd3b39b532dc1fd36dec1620d8e1210826/pkg/sql/logictest/logic.go#L1107)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

a bunch of these calls have the `/* initialize */` in the wrong place
</blockquote></details>

Done. TIL that gofmt will simplify `param a, param b, /* comment */ param c` to `param a, param b /* comment */, param c`. 


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45455#-:-M12aXV3-k66qUFpP8F7:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @tbg)



<!-- Sent from Reviewable.io -->

irfansharif commented:
Should this go into 20.1?

On Wed, Feb 26, 2020 at 5:24 PM Daniel Harrison <notifications@github.com>
wrote:

> *@danhhz* commented on this pull request.
>
> [image: :lgtm:]
> <https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67>
>
> *Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/45455#-:-M12aXV3-k66qUFpP8F7:bnfp4nl>*
> status: [image: :shipit:] complete! 1 of 0 LGTMs obtained (waiting on
> @otan <https://github.com/otan> and @tbg <https://github.com/tbg>)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/45455?email_source=notifications&email_token=ACQMN4QILFRNYBB3NZTAX63RE3T3DA5CNFSM4K4KT3U2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCXC57XY#pullrequestreview-365289439>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQMN4R7S57OTSB4Q3JRHQDRE3T3DANCNFSM4K4KT3UQ>
> .
>


-- 
Irfan Sharif, Staff Of Technical Member
*Cockroach Labs*

danhhz commented:
I vote yes, but curious what @tbg thinks.

There may even be an argument for sneaking in the RPC based version update system to the same in-memory state as what gossip updates, but not having anyone send those RPC. We should think through if that helps us in the 20.1->20.2 transition.

irfansharif committed:
settings/cluster: `git mv settings.go cluster_settings.go`

Reduce the number of packages and files named `settings`.

irfansharif committed:
clusterversion: significantly rework cluster version handling

We introduce clusterversion.Handle as a read only view to the
active cluster version and this binary's version details. We similarly
fold in the actual global cluster version setting into
pkg/clusterversion, and enforce all external usage to go through
clusterversion.Handle. We can also hide away external usage of the baked
in cluster.Binary{,MinimumSupported}Version constants. Instead we have
the implementation of clusterversion.Handle allow for tests to override
binary and minimum supported versions as needed.

Along the way we clean up Settings.Initialized, as it is not really
used. We document all the various "versions" in play ("binary version",
"minimum supported version", "active version") and change usage of what
was previously "serverVersion" to simply be "binaryVersion", because
that's what it is. We also clean up the Settings constructors into
Test/Non-Test types that differ in cluster version setting
initialization behaviour.

---

For reviewers: It's probably easiest to start with what
pkg/settings/cluster/cluster_settings.go looks like, then following into
pkg/clusterversion/clusterversion.go and then pkg/clusterversion/setting.go.

---

I still don't like the following detail about our pkg structure:

- pkg/settings/cluster depends on it's "parent" pkg, pkg/settings
- pkg/settings/cluster depends pkg/clusterversion, which in turn depends
  on pkg/settings

Naively, pkg/settings/cluster should be a top level package, but I'm not
making that change in this PR. For now I've renamed the settings.go file
to cluster_settings.go.

Release note: None

irfansharif commented:
@rohany: Mind rubberstamping `create_table.go`, last commit? Following up on https://github.com/cockroachdb/cockroach/pull/42922#pullrequestreview-366020508. Ignoring the discussions about whether or not `st` can be nil, I'll leave it someone else (you) to figure that out.

rohany commented on pkg/sql/create_table.go:
```diff
@@ -1210,7 +1210,7 @@ func MakeTableDesc(
 					return desc, invalidClusterForShardedIndexError
 				}
-				if version := cluster.Version.ActiveVersionOrEmpty(ctx, st); version == (clusterversion.ClusterVersion{}) ||
+				if version := st.Version.ActiveVersionOrEmpty(ctx); version == (clusterversion.ClusterVersion{}) ||
```
i think this should be !=

irfansharif replied:
Meant to throw a `!` in there, but that aside, I do think we want to error out if the version is not initialized, no?

irfansharif replied:
(I don't really know what's going on with `version != (clusterversion.ClusterVersion{})` above, we're not really handling the uninitialized version properly I think.)

rohany replied:
I'm going to take a look at this soon (#45519), so i wouldn't sweat about this. When i can move it to `ActiveVersion` it will error our if the version is not initialized.

irfansharif commented:
@tbg, gentle ping here. I've had to rebase it multiple times to stay afloat of master surprisingly. It's getting a bit slow to try and make sure each commit is functional on its own, so I'll just give up entirely until it's all reviewed.

tbg reviewed (approved):
I have only reviewed this superficially, but the general direction looks good. There are no semantic changes intended in any of the commits, right?

irfansharif commented:
yep, there isn't. Just the non-controversial parts of https://github.com/cockroachdb/cockroach/pull/44249 and then some. TFTRs.

irfansharif commented:
bors r+

tbg commented:
bors r+

tbg commented:
bors r+

irfansharif commented:
bors r+

irfansharif commented:
bors retry

irfansharif commented:
bors r+

On Sun, Mar 1, 2020, 1:53 AM craig[bot] <notifications@github.com> wrote:

> Timed out
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/45455?email_source=notifications&email_token=ACQMN4TPGSEZ2JMGYPQ2GV3RFIA47A5CNFSM4K4KT3U2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENMW4PQ#issuecomment-593063486>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQMN4X65WIEETCXRWQXG2DRFIA47ANCNFSM4K4KT3UQ>
> .
>

rohany commented:
what is going on with bors rn

tbg commented:
I thought it might be this PR, but I might be wrong

irfansharif commented:
[This](https://teamcity.cockroachdb.com/viewType.html?buildTypeId=Cockroach_UnitTests&branch_Cockroach=staging&tab=buildTypeStatusDiv) and [this](https://teamcity.cockroachdb.com/viewLog.html?buildId=1777100&buildTypeId=Cockroach_UnitTests) are the two most recent builds that were successful, but that bors timed out on.

irfansharif commented:
<img width="1557" alt="image" src="https://user-images.githubusercontent.com/10536690/75691033-c1297180-5c71-11ea-8e1e-ef6c8446c3bc.png">

They are taking quite a bit longer than other builds however, specifically for testrace. I don't quite understand why.

<img width="1561" alt="image" src="https://user-images.githubusercontent.com/10536690/75691123-d30b1480-5c71-11ea-971f-be37b53111c9.png">

There's no one lagging test there either.

<img width="1575" alt="image" src="https://user-images.githubusercontent.com/10536690/75691271-09489400-5c72-11ea-92cf-bae2598ff04c.png">

danhhz commented:
testrace taking longer is probably the "maybe stress new tests" logic

danhhz commented:
Nope. "Maybe stressrace pull request (17s)"

danhhz commented:
Are they taking longer for testrace, though? It looking to be taking a long time on all the recent master builds https://teamcity.cockroachdb.com/viewType.html?buildTypeId=Cockroach_UnitTests_Testrace&branch_Cockroach_UnitTests=%3Cdefault%3E&tab=buildTypeStatusDiv

irfansharif commented:
bors timeout bumped in https://github.com/cockroachdb/cockroach/pull/45590. bors r+

irfansharif commented:
bors r+
